provisioner/aws-manager.js

```diff
@@ -0,0 +1,696 @@
+    return Promise.all(that.managedRegions().filter(function(region) {
+      return !!res[region].KeyPairs[0];
+    }).map(function(region) {
+      debug('deleting key %s in %s', keyName, region);
+      return that.ec2.deleteKeyPair.inRegion(region, {
+        KeyName: keyName,
+      });
+    }));
+  });
+
+  p = p.then(function() {
+    that.__knownKeyPairs = that.__knownKeyPairs.filter(function(knownKeyPair) {
+      return knownKeyPair !== workerName;
+    });
+  });
+
+  return p;
+
+};
+
+
+/**
+ * Rouge Killer.  A rouge is an instance that has a KeyPair name
+ * that belongs to this provisioner but is not present in the list
+ * of workerNames provided.  We can also use this to shut down all
+ * instances of everything if we just pass an empty list of workers
+ * which will say to this function that all workerTypes are rouge.
+ * Sneaky, huh?
+ */
+AwsManager.prototype.rougeKiller = function(configuredWorkers) {
+  assert(configuredWorkers);
+  var that = this;
+  var workersInState = this.knownWorkerTypes();
+  var rouge = [];
+  workersInState.filter(function(name) {
+    return !configuredWorkers.includes(name) ;
+  }).forEach(function(name) {
+    debug('killing rouge instances for type %s', name);
+    rouge.push(that.deleteKeyPair(name));
+    rouge.push(that.killByName(name));
+  });
+
+  return Promise.all(rouge);
+};
+
+
+/**
+ * Kill all instances in all regions of a given workerName
+ */
+AwsManager.prototype.killByName = function(name) {
+  var deaths = [];
+  var that = this;
+
+  that.managedRegions().forEach(function(region) {
+    var apiState = that.getApi(region, name) || {};
+    var internalState = that.getInternal(region, name) || {};
+
+    var instances = [];
+    var requests = [];
+
+    [apiState, internalState].forEach(function(state) {
+      if (state.running) {
+        Array.prototype.push.apply(instances, state.running.map(function(r) {
+          return r.InstanceId;
+        }));
+      }
+
+      if (state.pending) {
+        Array.prototype.push.apply(instances, state.pending.map(function(r) {
+          return r.InstanceId;
+        }));
+      }
+
+      if (state.spotReq) {
+        Array.prototype.push.apply(requests, state.spotReq.map(function(r) {
+          // Remember that internal state is wrapped with some meta data!
+          if (r.request) {
+            return r.request.SpotInstanceRequestId;
+          } else {
+            return r.SpotInstanceRequestId;
+          }
+        }));
+      }
+    });
+
+    deaths.push(that.killCancel(region, instances, requests));
+  });
+
+  return Promise.all(deaths);
+};
+
+
+/**
+ * Kill instances and cancel spot requests
+ */
+AwsManager.prototype.killCancel = function(region, instances, requests) {
+  assert(instances || requests);
+  var that = this;
+
+  var promises = [];
+  var i = instances || [];
+  var r = requests || [];
+
+  if (i.length > 0) {
+    debug('killing instances %s in %s', JSON.stringify(i), region);
```

Comment:
Also not super important, but you might want to know that, both `debug` and `console.log` supports `%j` for json:
```js
debug('killing instances %j in %s', i, region);
```
So no need to stringify :)